eras/conway/impl/src/Cardano/Ledger/Conway/Governance/Proposals.hs

```diff
@@ -1,132 +1,424 @@
+import Data.Aeson (KeyValue ((.=)), ToJSON (..), object, pairs)
 import Data.Default.Class (Default (..))
+import Data.Foldable (foldl')
 import Data.Functor (($>))
+import Data.Kind (Type)
 import qualified Data.Map as Map
 import Data.Map.Strict (Map)
+import Data.Maybe (isNothing)
 import qualified Data.OMap.Strict as OMap
+import Data.Sequence (Seq)
 import Data.Sequence.Strict (StrictSeq (..))
 import Data.Set (Set)
 import qualified Data.Set as Set
 import GHC.Generics (Generic)
 import Lens.Micro
 import NoThunks.Class (NoThunks)
 
-newtype Proposals era
-  = Proposals
-      (OMap.OMap (GovActionId (EraCrypto era)) (GovActionState era))
-  deriving newtype (Show, Eq)
-  deriving stock (Generic)
+data PRoot a = PRoot
+  { prRoot :: !(Maybe a)
+  , prChildren :: !(Set a)
+  }
+  deriving (Show, Eq, Ord, Generic, NoThunks, NFData, Default)
+
+prRootL :: Lens' (PRoot a) (Maybe a)
+prRootL = lens prRoot $ \x y -> x {prRoot = y}
+
+prChildrenL :: Lens' (PRoot a) (Set a)
+prChildrenL = lens prChildren $ \x y -> x {prChildren = y}
+
+data PNode a = PNode
+  { pnParent :: !(Maybe a)
+  , pnChildren :: !(Set a)
+  }
+  deriving (Show, Eq, Ord, Generic, NoThunks, NFData, Default)
+
+pnChildrenL :: Lens' (PNode a) (Set a)
+pnChildrenL = lens pnChildren $ \x y -> x {pnChildren = y}
+
+newtype PHierarchy a = PHierarchy
+  { unPHierarchy :: Map a (PNode a)
+  }
+  deriving stock (Show, Eq, Ord, Generic)
+  deriving newtype (NoThunks, NFData, Default)
+
+pHierarchyNTL :: Lens' (PHierarchy a) (Map a (PNode a)) -- FIXME: what a terrible name!
+pHierarchyNTL = lens unPHierarchy $ \x y -> x {unPHierarchy = y}
+
+data PForest (f :: Type -> Type) era = PForest
+  { pfPParamUpdate :: !(f (GovPurposeId 'PParamUpdatePurpose era))
+  , pfHardFork :: !(f (GovPurposeId 'HardForkPurpose era))
+  , pfCommittee :: !(f (GovPurposeId 'CommitteePurpose era))
+  , pfConstitution :: !(f (GovPurposeId 'ConstitutionPurpose era))
+  }
+
+deriving instance
+  ( Show (f (GovPurposeId 'PParamUpdatePurpose era))
+  , Show (f (GovPurposeId 'HardForkPurpose era))
+  , Show (f (GovPurposeId 'CommitteePurpose era))
+  , Show (f (GovPurposeId 'ConstitutionPurpose era))
+  ) =>
+  Show (PForest f era)
+
+pfPParamUpdateL :: Lens' (PForest f era) (f (GovPurposeId 'PParamUpdatePurpose era))
+pfPParamUpdateL = lens pfPParamUpdate $ \x y -> x {pfPParamUpdate = y}
+
+pfHardForkL :: Lens' (PForest f era) (f (GovPurposeId 'HardForkPurpose era))
+pfHardForkL = lens pfHardFork $ \x y -> x {pfHardFork = y}
+
+pfCommitteeL :: Lens' (PForest f era) (f (GovPurposeId 'CommitteePurpose era))
+pfCommitteeL = lens pfCommittee $ \x y -> x {pfCommittee = y}
+
+pfConstitutionL :: Lens' (PForest f era) (f (GovPurposeId 'ConstitutionPurpose era))
+pfConstitutionL = lens pfConstitution $ \x y -> x {pfConstitution = y}
+
+pfrPParamUpdateL :: Lens' (PForest PRoot era) (PRoot (GovPurposeId 'PParamUpdatePurpose era))
+pfrPParamUpdateL = lens pfPParamUpdate $ \x y -> x {pfPParamUpdate = y}
```

Comment:
Couldn't we define these ones in terms of the `pf` ones,  in order to duplication? 
```
pfrPParamUpdateL = pfPParamUpdateL
..
pfhPParamUpdateL = pfPParamUpdateL
..
```
```

Reply:
You mean like this: `pfrPParamUpdateL = pfPParamUpdateL`?

Reply:
Yeah, and same for the hierarchy , `pfhPParamUpdateL = pfPParamUpdateL`

Reply:
It doesn't make much difference anyway..

Reply:
It definitely helps reduce visual noise :+1: I will try this. Thanks!